src/ident/AclIdent.cc

```diff
@@ -87,36 +85,56 @@ ACLIdent::empty () const
     return data->empty();
 }
 
+bool
+ACLIdent::ShouldStartLookup(const ACLFilledChecklist &ch)
+{
+    if (const auto c = ch.acceptedConnection()) {
+        const auto &clientConnection = *c;
+
+        // check this before checking isOpen() for more informative debugging
+        if (clientConnection.identLookup) {
+            debugs(82, 7, "no; already attempted");
+            return false;
+        }
+
+        if (!clientConnection.isOpen()) {
+            debugs(82, 5, "no; client connection closed: " << clientConnection.id);
+            return false;
+        }
+
+        debugs(82, 7, "yes for " << clientConnection.id);
+        return true;
+    }
+
+    debugs(82, 5, "no; not a client-associated transaction or insufficient client info");
+    return false;
+}
+
 void
 ACLIdent::StartLookup(ACLFilledChecklist &cl, const Acl::Node &)
 {
-    const ConnStateData *conn = cl.conn();
-    // check that ACLIdent::match() tested this lookup precondition
-    assert(conn && Comm::IsConnOpen(conn->clientConnection));
+    const auto clientConnection = cl.acceptedConnection();
+    // caller should use ShouldStartLookup() to check these preconditions
+    assert(clientConnection);
+    assert(Comm::IsConnOpen(clientConnection)); // TODO: Remove as unused/unnecessary
```

Comment:
Don't we need also to assert(!clientConnection.identLookup)?

Reply:
We could, but I would not: IIRC, that invariant is not used by (and is not technically necessary for) the Ident transaction code initiated below these assertions. That Ident transaction does not even call setIdent()/updateIdent() -- it does not know where the lookup outcome will be stored!

FWIW, I have added "Remove as unused/unnecessary" TODO for the other assertion for similar reasons -- we should not assert unused/unnecessary/inapplicable conditions, even if they are likely to be true in current code.

Yes, Ident transaction code may lookup identity of an already identified user, but some future hypothetical caller may have a reason for doing that (e.g., an identity cache refresh for long-lived connections?).

If we do add such an assertion now, then we must also add a ShouldStartLookup() check to ConnStateData::whenClientIpKnown(). Perhaps we should do that even if we do not add that assertion. My primary worry there is that it might trigger requests to merge ShouldStartLookup() and ACLIdent::StartLookup() calls as far as public API is concerned -- it may be difficult to merge them correctly (and to convince others that it is too difficult). Adding this assert may also trigger requests to move updateIdent() calls from callbacks to Ident transaction code. Again, such requests may not be wrong in principle, but they will increase this PR complexity further. I was hoping that those improvements can be done later if somebody wants to polish this rarely used code further.

I do realize that the code above implies that ShouldStartLookup() should be called. I think that misleading C++ comment should be removed (instead of adding an assertion). I was probably trying to fix inappropriate match() reference in the official comment, but my replacement was problematic.

Please let me know if any of my assumptions are false. If you agree with my reasoning, let's remove "caller should use..." comment and add a ShouldStartLookup() check to ConnStateData::whenClientIpKnown().

Reply:
>  let's remove "caller should use..." comment and add a ShouldStartLookup() check

Agreed, done at 7e707a5. Should we also remove the Comm::IsConnOpen(...) assertion?

Reply:
> Should we also remove the Comm::IsConnOpen(...) assertion?

Unless this removal is requested by others, I would keep this assertion (and the corresponding check in ShouldStartLookup()) in this PR to emphasize that we are not changing that precondition. My primary concern here is whether Ident lookup should be done for already closed connections (from general Ident principles unrelated to Squid Ident implementation). I do not know enough about these Ident services to address that concern.